tianshou/data/batch.py

```diff
@@ -593,23 +601,69 @@ def stack_(self,
                 if not issubclass(v.dtype.type, (np.bool_, np.number)):
                     v = v.astype(np.object)
                 self.__dict__[k] = v
-        keys_partial = reduce(set.symmetric_difference, keys_map)
+        keys_partial = set.union(*keys_map) - keys_shared
         _assert_type_keys(keys_partial)
         for k in keys_partial:
+            is_dict = False
+            value = None
             for i, e in enumerate(batches):
                 val = e.get(k, None)
                 if val is not None:
-                    try:
-                        self.__dict__[k][i] = val
-                    except KeyError:
-                        self.__dict__[k] = \
-                            _create_value(val, len(batches))
-                        self.__dict__[k][i] = val
+                    if isinstance(val, (dict, Batch)):
+                        is_dict = True
+                    else:  # np.ndarray / scalar or torch.Tensor
+                        value = val
+                    break
+            if is_dict:
+                self.__dict__[k] = Batch.stack(
+                    [e.get(k, Batch()) for e in batches], axis)
+            else:
+                if isinstance(value, np.ndarray) \
+                        or issubclass(value.__class__, np.generic) \
+                        or isinstance(value, Number):
+                    if isinstance(value, Number):
+                        value = np.asarray(value)
+                    arrs = []
+                    for i, e in enumerate(batches):
```

Comment:
What you are doing is almost duplicated for each if case, maybe you could find a way to factorize this. Duplicated blocks make it harder to maintain.